akka-stream/src/main/scala/akka/stream/impl/FoldResourceSinkAsync.scala

```
@@ -0,0 +1,162 @@
+import scala.util.control.NonFatal
+
+/**
+ * INTERNAL API
+ */
+@InternalApi private[akka] final class FoldResourceSinkAsync[T, S](
+  create: () ⇒ Future[S],
+  write:  (S, T) ⇒ Future[Unit],
+  close:  (S) ⇒ Future[Unit]) extends GraphStageWithMaterializedValue[SinkShape[T], Future[Done]] {
+  val in = Inlet[T]("FoldResourceSinkAsync.in")
+  override val shape = SinkShape(in)
+  override def initialAttributes: Attributes = DefaultAttributes.foldResourceSinkAsync
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, Future[Done]) = {
+    val promise = Promise[Done]()
+
+    val createLogic = new GraphStageLogic(shape) with InHandler {
+      lazy val decider = inheritedAttributes.mandatoryAttribute[SupervisionStrategy].decider
+
+      private implicit def ec = materializer.executionContext
+
+      private var blockingStream: Option[S] = None
+
+      private var writeFail: Option[Throwable] = None
+
+      private val createdCallback = getAsyncCallback[Try[S]] {
+        case Success(resource) ⇒
+          blockingStream = Some(resource)
+          pull(in)
+
+        case Failure(t) ⇒ failStage(t)
+      }.invokeWithFeedback _
+
+      private val errorHandler: PartialFunction[Throwable, Unit] = {
+        case NonFatal(ex) ⇒ decider(ex) match {
+          case Supervision.Stop ⇒
+            writeFail = Some(ex)
+            failStage(ex)
+          case Supervision.Restart ⇒ restartResource()
+          case Supervision.Resume  ⇒ pull(in)
+        }
+      }
+
+      private val pushCallback = getAsyncCallback[Try[Unit]] {
+        case Success(_) ⇒ pull(in)
+        case Failure(t) ⇒ errorHandler(t)
+      }.invoke _
+
+      override def preStart(): Unit = createResource()
+
+      override def onPush(): Unit =
+        blockingStream match {
+          case Some(resource) ⇒
+            try {
+              write(resource, grab(in)).onComplete(pushCallback)(sameThreadExecutionContext)
```

If the write fails, but the stage stops while write is happening, that exception is lost.